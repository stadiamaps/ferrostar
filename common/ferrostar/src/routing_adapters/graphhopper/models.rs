use serde::{Deserialize};
use alloc::collections::BTreeMap as HashMap;
use crate::models::{ GeographicCoordinate };

#[derive(Deserialize, Debug)]
pub struct GraphHopperRouteResponse {
    pub paths: Vec<GraphHopperPath>,
    pub message: Option<String>
}

#[derive(Deserialize, Debug)]
#[serde(untagged)]
pub enum DetailEntryValue {
    Int(i32),
    Str(String),
    Bool(bool),
    Float(f64),
}

#[derive(Deserialize, Debug)]
#[serde(from = "Vec<serde_json::Value>")]
pub struct DetailEntry {
    pub start_index: usize,
    pub end_index: usize,
    pub value: Option<DetailEntryValue>,
}

impl From<Vec<serde_json::Value>> for DetailEntry {
    fn from(vec: Vec<serde_json::Value>) -> Self {
        let start_index = vec[0].as_i64().unwrap_or_default() as usize;
        let end_index = vec[1].as_i64().unwrap_or_default() as usize;

        let value = match &vec[2] {
            serde_json::Value::String(s) => Some(DetailEntryValue::Str(s.clone())),
            serde_json::Value::Number(n) => {
                if n.is_i64() {
                    Some(DetailEntryValue::Int(n.as_i64().unwrap() as i32))
                } else {
                    Some(DetailEntryValue::Float(n.as_f64().unwrap()))
                }
            },
            serde_json::Value::Bool(b) => Some(DetailEntryValue::Bool(*b)),
            serde_json::Value::Null => None,
            _ => None,
        };

        DetailEntry { start_index, end_index, value }
    }
}

#[derive(Deserialize, Debug)]
pub struct GraphHopperPath {
    pub distance: f64,
    pub time: f64,
    pub bbox: Vec<f64>,
    pub instructions: Vec<GraphHopperInstruction>,
    pub details: HashMap<String, Vec<DetailEntry>>,
    pub points: String, // encoded polyline
    pub points_encoded:	bool,
    pub points_encoded_multiplier: f64,
}

#[derive(Deserialize, Debug)]
pub struct GraphHopperInstruction {
    pub distance: f64,
    pub time: f64,
    pub heading: Option<f64>,
    pub exit_number: Option<u32>,
    pub turn_angle: Option<f64>,
    pub sign: i32,
    pub interval: Vec<usize>,
    pub text: String,
    pub street_ref: Option<String>,
    pub street_name: String,
}

// temporary structure until we understand internal maxspeed handling
pub struct MaxSpeedEntry {
    pub geometry: Vec<GeographicCoordinate>,
    pub speed_limit: Option<f64>,
    pub unit: String,
}
