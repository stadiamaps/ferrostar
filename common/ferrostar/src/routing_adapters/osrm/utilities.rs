use super::models::AnyAnnotation;
use crate::models::AnyAnnotationValue;
use crate::routing_adapters::error::ParsingError;
use serde_json::Value;
use std::collections::HashMap;

/// Gets a slice of the route's annotations array.
///
/// Returns a [`ParsingError`] if the annotations are not present or the slice is out of bounds.
pub(crate) fn get_annotation_slice(
    annotations: Option<Vec<AnyAnnotationValue>>,
    start_index: usize,
    end_index: usize,
) -> Result<Vec<AnyAnnotationValue>, ParsingError> {
    annotations
        .ok_or(ParsingError::MalformedAnnotations {
            error: "No annotations".to_string(),
        })?
        .get(start_index..end_index)
        .ok_or(ParsingError::MalformedAnnotations {
            error: "Annotations slice index out of bounds ({start_index}..{end_index})".to_string(),
        })
        .map(<[AnyAnnotationValue]>::to_vec)
}

/// Converts the the OSRM-style annotation object consisting of separate arrays
/// to a single vector of parsed objects (one for each coordinate pair).
pub(crate) fn zip_annotations(annotation: AnyAnnotation) -> Vec<AnyAnnotationValue> {
    let source: HashMap<String, Vec<Value>> = annotation.values;

    // Get the length of the array (assumed to be the same for all annotations)
    let length = source.values().next().map_or(0, Vec::len);

    return (0..length)
        .map(|i| {
            source
                .iter()
                .filter_map(|(key, values)| {
                    // Values is the vector at a specific key in the original annotations object.
                    values
                        .get(i) // For each key, get the value at the index i.
                        .map(|v| (key.clone(), v.clone())) // Convert the key and value to a tuple.
                })
                .collect::<HashMap<String, Value>>() // Collect the key-value pairs into a hashmap.
        })
        .map(|value| AnyAnnotationValue { value })
        .collect::<Vec<AnyAnnotationValue>>();
}

#[cfg(test)]
mod test {

    use serde_json::Map;

    use super::*;

    #[test]
    fn test_zip_annotation() {
        let json_str = r#"{
            "distance": [1.2, 2.24, 3.24],
            "duration": [4, 5, 6],
            "speed": [10, 11, 12],
            "max_speed": [{
              "speed": 56,
              "unit": "km/h"
            }, {
              "speed": 12,
              "unit": "mi/h"
            }, {
              "unknown": true
            }],
            "traffic": ["bad", "ok", "good"],
            "construction": [null, true, null]
        }"#;

        let json_value: Map<String, Value> = serde_json::from_str(json_str).unwrap();
        let values: HashMap<String, Vec<Value>> = json_value
            .iter()
            .map(|(k, v)| (k.to_string(), v.as_array().unwrap().clone()))
            .collect();

        // Construct the annotation object.
        let annotation = AnyAnnotation { values };
        let result = zip_annotations(annotation);

        insta::with_settings!({sort_maps => true}, {
            insta::assert_yaml_snapshot!(result);
        });
    }
}
