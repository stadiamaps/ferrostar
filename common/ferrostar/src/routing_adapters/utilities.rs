use polyline::decode_polyline;

use crate::models::GeographicCoordinate;

use super::error::ParsingError;

/// Parse a polyline-encoded geometry string into a list of geographic coordinates.
/// If the polyline cannot be decoded, a [`ParsingError`] results.
pub fn get_coordinates_from_geometry(
    geometry: &str,
    polyline_precision: u32,
) -> Result<Vec<GeographicCoordinate>, ParsingError> {
    let linestring = decode_polyline(geometry, polyline_precision).map_err(|error| {
        ParsingError::InvalidGeometry {
            error: error.to_string(),
        }
    })?;

    // TODO: Trait for this common pattern?
    let linestring = linestring
        .coords()
        .map(|coord| GeographicCoordinate::from(*coord))
        .collect();

    Ok(linestring)
}
