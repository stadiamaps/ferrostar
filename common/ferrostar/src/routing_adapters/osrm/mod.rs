//! Response parsing for OSRM-compatible JSON (including Stadia Maps, Valhalla, Mapbox, etc.).

pub(crate) mod models;
pub mod utilities;

use super::RouteResponseParser;
use crate::models::{
    AnyAnnotationValue, GeographicCoordinate, Incident, LaneInfo, RouteStep, SpokenInstruction,
    VisualInstruction, VisualInstructionContent, Waypoint, WaypointKind,
};
use crate::routing_adapters::utilities::get_coordinates_from_geometry;
use crate::routing_adapters::{
    osrm::models::{
        Route as OsrmRoute, RouteResponse, RouteStep as OsrmRouteStep, Waypoint as OsrmWaypoint,
    },
    ParsingError, Route,
};
#[cfg(all(not(feature = "std"), feature = "alloc"))]
use alloc::{string::ToString, vec, vec::Vec};
use geo::BoundingRect;
use models::BannerContent;
use polyline::decode_polyline;
use utilities::get_annotation_slice;
use uuid::Uuid;

/// A response parser for OSRM-compatible routing backends.
///
/// The parser is NOT limited to only the standard OSRM format; many Valhalla/Mapbox tags are also
/// parsed and are included in the final route.
#[derive(Debug)]
pub struct OsrmResponseParser {
    polyline_precision: u32,
}

impl OsrmResponseParser {
    pub fn new(polyline_precision: u32) -> Self {
        Self { polyline_precision }
    }
}

impl RouteResponseParser for OsrmResponseParser {
    fn parse_response(&self, response: Vec<u8>) -> Result<Vec<Route>, ParsingError> {
        let res: RouteResponse = serde_json::from_slice(&response)?;

        if res.code == "Ok" {
            res.routes
                .iter()
                .map(|route| Route::from_osrm(route, &res.waypoints, self.polyline_precision))
                .collect::<Result<Vec<_>, _>>()
        } else {
            Err(ParsingError::InvalidStatusCode { code: res.code })
        }
    }
}

impl Route {
    pub fn from_osrm(
        route: &OsrmRoute,
        waypoints: &[OsrmWaypoint],
        polyline_precision: u32,
    ) -> Result<Self, ParsingError> {
        let via_waypoint_indices: Vec<_> = route
            .legs
            .iter()
            .flat_map(|leg| leg.via_waypoints.iter().map(|via| via.waypoint_index))
            .collect();

        let waypoints: Vec<_> = waypoints
            .iter()
            .enumerate()
            .map(|(idx, waypoint)| Waypoint {
                coordinate: GeographicCoordinate {
                    lat: waypoint.location.latitude(),
                    lng: waypoint.location.longitude(),
                },
                kind: if via_waypoint_indices.contains(&idx) {
                    WaypointKind::Via
                } else {
                    WaypointKind::Break
                },
            })
            .collect();

        let linestring = decode_polyline(&route.geometry, polyline_precision).map_err(|error| {
            ParsingError::InvalidGeometry {
                error: error.to_string(),
            }
        })?;
        if let Some(bbox) = linestring.bounding_rect() {
            let geometry: Vec<GeographicCoordinate> = linestring
                .coords()
                .map(|coord| GeographicCoordinate::from(*coord))
                .collect();

            let steps = route
                .legs
                .iter()
                .flat_map(|leg| {
                    // Converts all single value annotation vectors into a single vector witih a value object.
                    let annotations = leg
                        .annotation
                        .as_ref()
                        .map(|leg_annotation| utilities::zip_annotations(leg_annotation.clone()));

                    // Convert all incidents into a vector of Incident objects.
                    let incident_items = leg
                        .incidents
                        .iter()
                        .map(Incident::from)
                        .collect::<Vec<Incident>>();

                    // Index for the annotations slice
                    let mut start_index: usize = 0;

                    leg.steps.iter().map(move |step| {
                        let step_geometry =
                            get_coordinates_from_geometry(&step.geometry, polyline_precision)?;

                        // Slice the annotations for the current step.
                        // The annotations array represents segments between coordinates.
                        //
                        // 1. Annotations should never repeat.
                        // 2. Each step has one less annotation than coordinate.
                        // 3. The last step never has annotations as it's two of the route's last coordinate (duplicate).
                        let step_index_len = step_geometry.len() - 1_usize;
                        let end_index = start_index + step_index_len;

                        let annotation_slice =
                            get_annotation_slice(annotations.clone(), start_index, end_index).ok();

                        let relevant_incidents_slice = incident_items
                            .iter()
                            .filter(|incident| {
                                let incident_start = incident.geometry_index_start as usize;

                                match incident.geometry_index_end {
                                    Some(end) => {
                                        let incident_end = end as usize;
                                        incident_start >= start_index && incident_end <= end_index
                                    }
                                    None => {
                                        incident_start >= start_index && incident_start <= end_index
                                    }
                                }
                            })
                            .map(|incident| {
                                let mut adjusted_incident = incident.clone();
                                if adjusted_incident.geometry_index_start - start_index as u64 > 0 {
                                    adjusted_incident.geometry_index_start -= start_index as u64;
                                } else {
                                    adjusted_incident.geometry_index_start = 0;
                                }

                                if let Some(end) = adjusted_incident.geometry_index_end {
                                    let adjusted_end = end - start_index as u64;
                                    adjusted_incident.geometry_index_end =
                                        Some(if adjusted_end > end_index as u64 {
                                            end_index as u64
                                        } else {
                                            adjusted_end
                                        });
                                }
                                adjusted_incident
                            })
                            .collect::<Vec<Incident>>();

                        start_index = end_index;

                        RouteStep::from_osrm_and_geom(
                            step,
                            step_geometry,
                            annotation_slice,
                            relevant_incidents_slice,
                        )
                    })
                })
                .collect::<Result<Vec<_>, _>>()?;

            Ok(Route {
                geometry,
                bbox: bbox.into(),
                distance: route.distance,
                waypoints: waypoints.clone(),
                steps,
            })
        } else {
            Err(ParsingError::InvalidGeometry {
                error: "Bounding box could not be calculated".to_string(),
            })
        }
    }
}

impl RouteStep {
    fn extract_exit_numbers(banner_content: &BannerContent) -> Vec<String> {
        banner_content
            .components
            .iter()
            .filter(|component| component.component_type.as_deref() == Some("exit-number"))
            .filter_map(|component| component.text.clone())
            .collect()
    }

    fn from_osrm_and_geom(
        value: &OsrmRouteStep,
        geometry: Vec<GeographicCoordinate>,
        annotations: Option<Vec<AnyAnnotationValue>>,
        incidents: Vec<Incident>,
    ) -> Result<Self, ParsingError> {
        let visual_instructions = value
            .banner_instructions
            .iter()
            .map(|banner| VisualInstruction {
                primary_content: VisualInstructionContent {
                    text: banner.primary.text.clone(),
                    maneuver_type: banner.primary.maneuver_type,
                    maneuver_modifier: banner.primary.maneuver_modifier,
                    roundabout_exit_degrees: banner.primary.roundabout_exit_degrees,
                    lane_info: None,
                    exit_numbers: Self::extract_exit_numbers(&banner.primary),
                },
                secondary_content: banner.secondary.as_ref().map(|secondary| {
                    VisualInstructionContent {
                        text: secondary.text.clone(),
                        maneuver_type: secondary.maneuver_type,
                        maneuver_modifier: secondary.maneuver_modifier,
                        roundabout_exit_degrees: banner.primary.roundabout_exit_degrees,
                        lane_info: None,
                        exit_numbers: Self::extract_exit_numbers(&secondary),
                    }
                }),
                sub_content: banner.sub.as_ref().map(|sub| VisualInstructionContent {
                    text: sub.text.clone(),
                    maneuver_type: sub.maneuver_type,
                    maneuver_modifier: sub.maneuver_modifier,
                    roundabout_exit_degrees: sub.roundabout_exit_degrees,
                    lane_info: {
                        let lane_infos: Vec<LaneInfo> = sub
                            .components
                            .iter()
                            .filter(|component| component.component_type.as_deref() == Some("lane"))
                            .map(|component| LaneInfo {
                                active: component.active.unwrap_or(false),
                                directions: component.directions.clone().unwrap_or_default(),
                                active_direction: component.active_direction.clone(),
                            })
                            .collect();

                        if lane_infos.is_empty() {
                            None
                        } else {
                            Some(lane_infos)
                        }
                    },
                    exit_numbers: Self::extract_exit_numbers(&sub),
                }),
                trigger_distance_before_maneuver: banner.distance_along_geometry,
            })
            .collect();

        let spoken_instructions = value
            .voice_instructions
            .iter()
            .map(|instruction| SpokenInstruction {
                text: instruction.announcement.clone(),
                ssml: instruction.ssml_announcement.clone(),
                trigger_distance_before_maneuver: instruction.distance_along_geometry,
                utterance_id: Uuid::new_v4(),
            })
            .collect();

        // Convert the annotations to a vector of json strings.
        // This allows us to safely pass the RouteStep through the FFI boundary.
        // The host platform can then parse an arbitrary annotation object.
        let annotations_as_strings: Option<Vec<String>> = annotations.map(|annotations_vec| {
            annotations_vec
                .iter()
                .map(|annotation| serde_json::to_string(annotation).unwrap())
                .collect()
        });

        let exits = match value.exits.clone() {
            Some(exit_text) => exit_text.split(';').map(|s| s.trim().to_string()).collect(),
            None => Vec::new(),
        };

        Ok(RouteStep {
            geometry,
            // TODO: Investigate using the haversine distance or geodesics to normalize.
            // Valhalla in particular is a bit nonstandard. See https://github.com/valhalla/valhalla/issues/1717
            distance: value.distance,
            duration: value.duration,
            road_name: value.name.clone(),
            exits,
            instruction: value.maneuver.get_instruction(),
            visual_instructions,
            spoken_instructions,
            annotations: annotations_as_strings,
            incidents,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    const STANDARD_OSRM_POLYLINE6_RESPONSE: &str =
        include_str!("fixtures/standard_osrm_polyline6_response.json");
    const VALHALLA_OSRM_RESPONSE: &str = include_str!("fixtures/valhalla_osrm_response.json");
    const VALHALLA_OSRM_RESPONSE_VIA_WAYS: &str =
        include_str!("fixtures/valhalla_osrm_response_via_ways.json");
    const VALHALLA_EXTENDED_OSRM_RESPONSE: &str =
        include_str!("fixtures/valhalla_extended_osrm_response.json");
    const VALHALLA_OSRM_RESPONSE_WITH_EXITS: &str =
        include_str!("fixtures/valhalla_osrm_response_with_exit_info.json");

    #[test]
    fn parse_standard_osrm() {
        let parser = OsrmResponseParser::new(6);
        let routes = parser
            .parse_response(STANDARD_OSRM_POLYLINE6_RESPONSE.into())
            .expect("Unable to parse OSRM response");
        insta::assert_yaml_snapshot!(routes);
    }

    #[test]
    fn parse_valhalla_osrm() {
        let parser = OsrmResponseParser::new(6);
        let routes = parser
            .parse_response(VALHALLA_OSRM_RESPONSE.into())
            .expect("Unable to parse Valhalla OSRM response");

        insta::assert_yaml_snapshot!(routes, {
            ".**.annotations" => "redacted annotations json strings vec"
        });
    }

    #[test]
    fn parse_valhalla_osrm_with_via_ways() {
        let parser = OsrmResponseParser::new(6);
        let routes = parser
            .parse_response(VALHALLA_OSRM_RESPONSE_VIA_WAYS.into())
            .expect("Unable to parse Valhalla OSRM response");

        insta::assert_yaml_snapshot!(routes, {
            ".**.annotations" => "redacted annotations json strings vec"
        });
    }

    #[test]
    fn parse_valhalla_asserting_annotation_lengths() {
        let parser = OsrmResponseParser::new(6);
        let routes = parser
            .parse_response(VALHALLA_OSRM_RESPONSE.into())
            .expect("Unable to parse Valhalla OSRM response");

        // Loop through every step and validate that the length of the annotations
        // matches the length of the geometry minus one. This is because each annotation
        // represents a segment between two coordinates.
        for (route_index, route) in routes.iter().enumerate() {
            for (step_index, step) in route.steps.iter().enumerate() {
                if step_index == route.steps.len() - 1 {
                    // The arrival step is 2 of the same coordinates.
                    // So annotations will be None.
                    assert_eq!(step.annotations, None);
                    continue;
                }

                let step = step.clone();
                let annotations = step.annotations.expect("No annotations");
                assert_eq!(
                    annotations.len(),
                    step.geometry.len() - 1,
                    "Route {}, Step {}",
                    route_index,
                    step_index
                );
            }
        }
    }

    #[test]
    fn parse_valhalla_asserting_sub_maneuvers() {
        let parser = OsrmResponseParser::new(6);
        let routes = parser
            .parse_response(VALHALLA_EXTENDED_OSRM_RESPONSE.into())
            .expect("Unable to parse Valhalla Extended OSRM response");

        // Collect all sub_contents into a vector
        let sub_contents: Vec<_> = routes
            .iter()
            .flat_map(|route| &route.steps)
            .filter_map(|step| {
                step.visual_instructions
                    .iter()
                    .find_map(|instruction: &VisualInstruction| instruction.sub_content.as_ref())
            })
            .collect();

        // Assert that there's exactly one sub maneuver instructions as is the case in the test data
        assert_eq!(
            sub_contents.len(),
            1,
            "Expected exactly one sub banner instructions"
        );

        if let Some(sub_content) = sub_contents.first() {
            // Ensure that there are 4 pieces of lane information in the sub banner instructions
            if let Some(lane_info) = &sub_content.lane_info {
                assert_eq!(lane_info.len(), 4);
            } else {
                panic!("Expected lane information, but could not find it");
            }
        } else {
            panic!("No sub banner instructions found in any of the steps")
        }
    }

    #[test]
    fn parse_osrm_with_exits() {
        let parser = OsrmResponseParser::new(6);
        let routes = parser
            .parse_response(VALHALLA_OSRM_RESPONSE_WITH_EXITS.into())
            .expect("Unable to parse OSRM response");

        insta::assert_yaml_snapshot!(routes, {
            ".**.annotations" => "redacted annotations json strings vec"
        });
    }
}
