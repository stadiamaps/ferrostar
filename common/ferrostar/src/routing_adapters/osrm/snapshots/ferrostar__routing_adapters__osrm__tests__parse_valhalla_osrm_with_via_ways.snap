---
source: ferrostar/src/routing_adapters/osrm/mod.rs
expression: routes
---
- geometry:
    - lat: 28.795656
      lng: -82.036056
    - lat: 28.795632
      lng: -82.03602
    - lat: 28.795446
      lng: -82.035862
    - lat: 28.795242
      lng: -82.035787
    - lat: 28.794865
      lng: -82.035633
    - lat: 28.794788
      lng: -82.035601
    - lat: 28.794687
      lng: -82.035548
    - lat: 28.794614
      lng: -82.03551
    - lat: 28.794567
      lng: -82.035518
    - lat: 28.794504
      lng: -82.035542
    - lat: 28.794436
      lng: -82.035511
    - lat: 28.794273
      lng: -82.035351
    - lat: 28.794139
      lng: -82.0352
    - lat: 28.793991
      lng: -82.035104
    - lat: 28.793934
      lng: -82.035044
    - lat: 28.793761
      lng: -82.034758
    - lat: 28.793597
      lng: -82.034492
    - lat: 28.793455
      lng: -82.034264
    - lat: 28.793318
      lng: -82.033963
    - lat: 28.79323
      lng: -82.033796
    - lat: 28.793138
      lng: -82.033645
    - lat: 28.793118
      lng: -82.033577
    - lat: 28.793107
      lng: -82.033531
    - lat: 28.793056
      lng: -82.033182
    - lat: 28.792997
      lng: -82.032977
    - lat: 28.792985
      lng: -82.032905
    - lat: 28.792979
      lng: -82.032845
    - lat: 28.792977
      lng: -82.032804
    - lat: 28.79296
      lng: -82.032524
    - lat: 28.792979
      lng: -82.032225
    - lat: 28.792998
      lng: -82.031982
    - lat: 28.793005
      lng: -82.031645
    - lat: 28.793011
      lng: -82.031423
    - lat: 28.793038
      lng: -82.031166
    - lat: 28.793149
      lng: -82.030774
    - lat: 28.793239
      lng: -82.030562
    - lat: 28.793445
      lng: -82.030225
    - lat: 28.79353
      lng: -82.030069
    - lat: 28.793602
      lng: -82.029863
    - lat: 28.793634
      lng: -82.029825
    - lat: 28.793661
      lng: -82.029777
    - lat: 28.793727
      lng: -82.029683
    - lat: 28.793881
      lng: -82.029487
    - lat: 28.793972
      lng: -82.029351
    - lat: 28.79407
      lng: -82.029159
    - lat: 28.79413
      lng: -82.028968
    - lat: 28.794172
      lng: -82.028687
    - lat: 28.794177
      lng: -82.028508
    - lat: 28.79416
      lng: -82.028333
    - lat: 28.794116
      lng: -82.028138
    - lat: 28.794097
      lng: -82.028035
    - lat: 28.794078
      lng: -82.027959
    - lat: 28.794075
      lng: -82.027904
    - lat: 28.794079
      lng: -82.02769
    - lat: 28.79408
      lng: -82.027554
    - lat: 28.794052
      lng: -82.027401
    - lat: 28.794058
      lng: -82.027272
    - lat: 28.794071
      lng: -82.027151
    - lat: 28.794099
      lng: -82.026985
    - lat: 28.794129
      lng: -82.02673
    - lat: 28.794118
      lng: -82.026562
    - lat: 28.794086
      lng: -82.026389
    - lat: 28.794032
      lng: -82.026212
    - lat: 28.793989
      lng: -82.026094
    - lat: 28.793908
      lng: -82.025904
    - lat: 28.793767
      lng: -82.02561
    - lat: 28.793575
      lng: -82.025339
    - lat: 28.793382
      lng: -82.025152
    - lat: 28.793051
      lng: -82.024889
    - lat: 28.792853
      lng: -82.024677
    - lat: 28.792666
      lng: -82.024502
    - lat: 28.792613
      lng: -82.024391
    - lat: 28.792553
      lng: -82.024229
    - lat: 28.792517
      lng: -82.024048
    - lat: 28.792452
      lng: -82.023868
    - lat: 28.792434
      lng: -82.02372
    - lat: 28.792409
      lng: -82.023507
    - lat: 28.792369
      lng: -82.023212
    - lat: 28.792307
      lng: -82.02297
    - lat: 28.792235
      lng: -82.022756
    - lat: 28.792102
      lng: -82.022403
    - lat: 28.791981
      lng: -82.022172
    - lat: 28.791823
      lng: -82.021943
    - lat: 28.791628
      lng: -82.02171
    - lat: 28.791463
      lng: -82.021494
    - lat: 28.791281
      lng: -82.021204
    - lat: 28.791179
      lng: -82.021008
    - lat: 28.791133
      lng: -82.020892
    - lat: 28.79109
      lng: -82.020755
    - lat: 28.791046
      lng: -82.020511
    - lat: 28.790976
      lng: -82.020256
    - lat: 28.790942
      lng: -82.020122
    - lat: 28.790931
      lng: -82.019895
    - lat: 28.790906
      lng: -82.019766
    - lat: 28.790841
      lng: -82.019569
    - lat: 28.790763
      lng: -82.019381
    - lat: 28.7906
      lng: -82.019078
    - lat: 28.790396
      lng: -82.01882
    - lat: 28.790264
      lng: -82.018582
    - lat: 28.790157
      lng: -82.018303
    - lat: 28.790106
      lng: -82.018021
  bbox:
    sw:
      lat: 28.790106
      lng: -82.036056
    ne:
      lat: 28.795656
      lng: -82.018021
  distance: 2089.442
  waypoints:
    - coordinate:
        lat: 28.795656
        lng: -82.036056
      kind: Break
    - coordinate:
        lat: 28.795446
        lng: -82.035862
      kind: Via
    - coordinate:
        lat: 28.790106
        lng: -82.018021
      kind: Break
  steps:
    - geometry:
        - lat: 28.795656
          lng: -82.036056
        - lat: 28.795632
          lng: -82.03602
        - lat: 28.795446
          lng: -82.035862
        - lat: 28.795242
          lng: -82.035787
        - lat: 28.794865
          lng: -82.035633
        - lat: 28.794788
          lng: -82.035601
        - lat: 28.794687
          lng: -82.035548
        - lat: 28.794614
          lng: -82.03551
        - lat: 28.794567
          lng: -82.035518
        - lat: 28.794504
          lng: -82.035542
        - lat: 28.794436
          lng: -82.035511
        - lat: 28.794273
          lng: -82.035351
        - lat: 28.794139
          lng: -82.0352
        - lat: 28.793991
          lng: -82.035104
        - lat: 28.793934
          lng: -82.035044
        - lat: 28.793761
          lng: -82.034758
        - lat: 28.793597
          lng: -82.034492
        - lat: 28.793455
          lng: -82.034264
        - lat: 28.793318
          lng: -82.033963
        - lat: 28.79323
          lng: -82.033796
        - lat: 28.793138
          lng: -82.033645
        - lat: 28.793118
          lng: -82.033577
        - lat: 28.793107
          lng: -82.033531
        - lat: 28.793056
          lng: -82.033182
        - lat: 28.792997
          lng: -82.032977
        - lat: 28.792985
          lng: -82.032905
        - lat: 28.792979
          lng: -82.032845
        - lat: 28.792977
          lng: -82.032804
        - lat: 28.79296
          lng: -82.032524
        - lat: 28.792979
          lng: -82.032225
        - lat: 28.792998
          lng: -82.031982
        - lat: 28.793005
          lng: -82.031645
        - lat: 28.793011
          lng: -82.031423
        - lat: 28.793038
          lng: -82.031166
        - lat: 28.793149
          lng: -82.030774
        - lat: 28.793239
          lng: -82.030562
        - lat: 28.793445
          lng: -82.030225
        - lat: 28.79353
          lng: -82.030069
        - lat: 28.793602
          lng: -82.029863
        - lat: 28.793634
          lng: -82.029825
        - lat: 28.793661
          lng: -82.029777
        - lat: 28.793727
          lng: -82.029683
        - lat: 28.793881
          lng: -82.029487
        - lat: 28.793972
          lng: -82.029351
        - lat: 28.79407
          lng: -82.029159
        - lat: 28.79413
          lng: -82.028968
        - lat: 28.794172
          lng: -82.028687
        - lat: 28.794177
          lng: -82.028508
        - lat: 28.79416
          lng: -82.028333
        - lat: 28.794116
          lng: -82.028138
        - lat: 28.794097
          lng: -82.028035
        - lat: 28.794078
          lng: -82.027959
        - lat: 28.794075
          lng: -82.027904
        - lat: 28.794079
          lng: -82.02769
        - lat: 28.79408
          lng: -82.027554
        - lat: 28.794052
          lng: -82.027401
        - lat: 28.794058
          lng: -82.027272
        - lat: 28.794071
          lng: -82.027151
        - lat: 28.794099
          lng: -82.026985
        - lat: 28.794129
          lng: -82.02673
        - lat: 28.794118
          lng: -82.026562
        - lat: 28.794086
          lng: -82.026389
        - lat: 28.794032
          lng: -82.026212
        - lat: 28.793989
          lng: -82.026094
        - lat: 28.793908
          lng: -82.025904
        - lat: 28.793767
          lng: -82.02561
        - lat: 28.793575
          lng: -82.025339
        - lat: 28.793382
          lng: -82.025152
        - lat: 28.793051
          lng: -82.024889
        - lat: 28.792853
          lng: -82.024677
        - lat: 28.792666
          lng: -82.024502
        - lat: 28.792613
          lng: -82.024391
        - lat: 28.792553
          lng: -82.024229
        - lat: 28.792517
          lng: -82.024048
        - lat: 28.792452
          lng: -82.023868
        - lat: 28.792434
          lng: -82.02372
        - lat: 28.792409
          lng: -82.023507
        - lat: 28.792369
          lng: -82.023212
        - lat: 28.792307
          lng: -82.02297
        - lat: 28.792235
          lng: -82.022756
        - lat: 28.792102
          lng: -82.022403
        - lat: 28.791981
          lng: -82.022172
        - lat: 28.791823
          lng: -82.021943
        - lat: 28.791628
          lng: -82.02171
        - lat: 28.791463
          lng: -82.021494
        - lat: 28.791281
          lng: -82.021204
        - lat: 28.791179
          lng: -82.021008
        - lat: 28.791133
          lng: -82.020892
        - lat: 28.79109
          lng: -82.020755
        - lat: 28.791046
          lng: -82.020511
        - lat: 28.790976
          lng: -82.020256
        - lat: 28.790942
          lng: -82.020122
        - lat: 28.790931
          lng: -82.019895
        - lat: 28.790906
          lng: -82.019766
        - lat: 28.790841
          lng: -82.019569
        - lat: 28.790763
          lng: -82.019381
        - lat: 28.7906
          lng: -82.019078
        - lat: 28.790396
          lng: -82.01882
        - lat: 28.790264
          lng: -82.018582
        - lat: 28.790157
          lng: -82.018303
        - lat: 28.790106
          lng: -82.018021
      distance: 2089.442
      duration: 301.262
      road_name: ""
      instruction: Drive southeast.
      visual_instructions:
        - primary_content:
            text: You have arrived at your destination.
            maneuver_type: arrive
            maneuver_modifier: ~
            roundabout_exit_degrees: ~
          secondary_content: ~
          trigger_distance_before_maneuver: 2089.442
      spoken_instructions:
        - text: Drive southeast.
          ssml: "<speak>Drive southeast.</speak>"
          trigger_distance_before_maneuver: 2089.442
        - text: "In 200 feet, You have arrived at your destination."
          ssml: "<speak>In 200 feet, You have arrived at your destination.</speak>"
          trigger_distance_before_maneuver: 60
      annotations: redacted annotations json strings vec
    - geometry:
        - lat: 28.790106
          lng: -82.018021
        - lat: 28.790106
          lng: -82.018021
      distance: 0
      duration: 0
      road_name: ""
      instruction: You have arrived at your destination.
      visual_instructions:
        - primary_content:
            text: You have arrived at your destination.
            maneuver_type: arrive
            maneuver_modifier: ~
            roundabout_exit_degrees: ~
          secondary_content: ~
          trigger_distance_before_maneuver: 0
      spoken_instructions: []
      annotations: redacted annotations json strings vec
